<a href="https://colab.research.google.com/github/yoonputer/leaning_machinelearning/blob/master/titanic_onehot_logisticregression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import seaborn as sns
df = sns.load_dataset('titanic')
df

survived  pclass     sex   age  ...  deck  embark_town  alive  alone
0           0       3    male  22.0  ...   NaN  Southampton     no  False
1           1       1  female  38.0  ...     C    Cherbourg    yes  False
2           1       3  female  26.0  ...   NaN  Southampton    yes   True
3           1       1  female  35.0  ...     C  Southampton    yes  False
4           0       3    male  35.0  ...   NaN  Southampton     no   True
..        ...     ...     ...   ...  ...   ...          ...    ...    ...
886         0       2    male  27.0  ...   NaN  Southampton     no   True
887         1       1  female  19.0  ...     B  Southampton    yes   True
888         0       3  female   NaN  ...   NaN  Southampton     no  False
889         1       1    male  26.0  ...     C    Cherbourg    yes   True
890         0       3    male  32.0  ...   NaN   Queenstown     no   True

[891 rows x 15 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.6+ KB


#### pclass, sex, sibsp, parch, embarked, class, who, adult_male, embark_town, alive, alone

not-null 수가 891개라(null이 x) 후보군

In [ ]:
import pandas as pd

In [ ]:
dfx = pd.get_dummies(df['sex'])
# sex : female -> 0, male -> 1
dfx

female  male
0         0     1
1         1     0
2         1     0
3         1     0
4         0     1
..      ...   ...
886       0     1
887       1     0
888       1     0
889       0     1
890       0     1

[891 rows x 2 columns]

In [ ]:
df = pd.concat([df,dfx],axis=1)
df

survived  pclass     sex   age  ...  alive  alone  female male
0           0       3    male  22.0  ...     no  False       0    1
1           1       1  female  38.0  ...    yes  False       1    0
2           1       3  female  26.0  ...    yes   True       1    0
3           1       1  female  35.0  ...    yes  False       1    0
4           0       3    male  35.0  ...     no   True       0    1
..        ...     ...     ...   ...  ...    ...    ...     ...  ...
886         0       2    male  27.0  ...     no   True       0    1
887         1       1  female  19.0  ...    yes   True       1    0
888         0       3  female   NaN  ...     no  False       1    0
889         1       1    male  26.0  ...    yes   True       0    1
890         0       3    male  32.0  ...     no   True       0    1

[891 rows x 17 columns]

In [ ]:
dfx = pd.get_dummies(df['pclass'],prefix='pclass')
df = pd.concat([df,dfx],axis=1)
df
#pd.get_dummies(df['class'],prefix='pclass',drop_first=True)
# y(survive) = ax(pclass) + 0

survived  pclass     sex   age  ...  male  pclass_1  pclass_2 pclass_3
0           0       3    male  22.0  ...     1         0         0        1
1           1       1  female  38.0  ...     0         1         0        0
2           1       3  female  26.0  ...     0         0         0        1
3           1       1  female  35.0  ...     0         1         0        0
4           0       3    male  35.0  ...     1         0         0        1
..        ...     ...     ...   ...  ...   ...       ...       ...      ...
886         0       2    male  27.0  ...     1         0         1        0
887         1       1  female  19.0  ...     0         1         0        0
888         0       3  female   NaN  ...     0         0         0        1
889         1       1    male  26.0  ...     1         1         0        0
890         0       3    male  32.0  ...     1         0         0        1

[891 rows x 20 columns]

female	male pclass_First	pclass_Second	pclass_Third

In [ ]:
df[['fare','female', 'male', 'pclass_1', 'pclass_2', 'pclass_3','survived']].head(5)

fare  female  male  pclass_1  pclass_2  pclass_3  survived
0   7.2500       0     1         0         0         1         0
1  71.2833       1     0         1         0         0         1
2   7.9250       1     0         0         0         1         1
3  53.1000       1     0         1         0         0         1
4   8.0500       0     1         0         0         1         0

In [ ]:
X = df[['fare','female', 'male', 'pclass_1', 'pclass_2', 'pclass_3']]
Y = df[['survived']]
X.shape, Y.shape

((891, 6), (891, 1))

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)
X.shape

(891, 6)

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X,Y)

In [ ]:
from sklearn.linear_model import LogisticRegression
logR = LogisticRegression()
logR.fit(x_train, y_train)

In [ ]:
logR.score(x_train, y_train)

0.7859281437125748